link/tracepoint.go

```diff
@@ -0,0 +1,41 @@
+package link
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/cilium/ebpf"
+)
+
+// Tracepoint returns a trace event for the given tracepoint group and name.
+// See /sys/kernel/debug/tracing/events to find available tracepoints. The
+// top-level directory is group, the event's subdirectory is name. Example:
+//
+//	Tracepoint("syscalls", "sys_enter_fork")
+//
+// Note that attaching eBPF programs to syscalls (sys_enter_*/sys_exit_*) is
+// only possible as of kernel 4.14 (commit cf5f5ce).
+//
+// Multiple calls with the same group and name are unnecessary. Tracepoints
+// can be reused to Attach() an arbitrary amount of eBPF programs. Closing a
+// TraceEvent created by Tracepoint will not disconnect attached perf events.
+func Tracepoint(group, name string) (*TraceEvent, error) {
```

Comment:
Can we create and expose a TraceEvent interface ? instead of a TraceEvent struct ? 

This would help for users writing unit tests.

Reply:
Could you elaborate on which behaviours you would like to test as a consumer of the package?

Reply:
I think for non-trivial BPF program deployments it would be important to test if attachments and close calls are correct. 

An example use case I am working on is creating and deleting uprobes for individual docker containers, which are constantly coming and going. My program needs to manage the lifecycle of these uprobes, Ideally I would be able to test the interaction with the `cilium/ebpf` library by just mocking interfaces, the alternative is introducing a shim between my code and the library. 
> The above example is irrelevant for kprobes, but it seems like the same API will be used for both program types.

Having said this returning an interface here is not a silver bullet for testing. My use case dynamically finds the binary that we need to attach too. Hence we need to dynamically call `link.Uprobe` (this is equivalent to dynamically figuring out symbol for `link.Kprobe`). `link.Uprobe` cannot be mocked and proper testing requires a shim.

Ideal solution for above is making `TraceEvent` an interface and making sure that the Uprobe API supports dynamic configuration of which binary file the uprobe is attached to.

Reply:
> making sure that the TraceEvent API supports dynamic configuration of which binary file the uprobe is attached to

I don't think the kernel supports moving an existing perf event to another symbol. That would require creating a new perf event. We're considering to implement `Link.Update(prog)` for perf events that would do this as a convenience method, but that wouldn't cover your use case.

That said, I've factored out `TraceEvent` completely last night. The API will be `Kprobe(symbol, *prog) (Link, error)` instead of a split create/attach stage, will push this in a bit.

Reply:
My 2c: we should avoid returning an interface in our API if possible, aka `accept interfaces, return structs`. If we do return an interface (that is not from the stdlib) we should make it non-implementable for users, like `Link`. Otherwise our hands are tied in terms of API evolution.

@etherandrius does anything prevent you from defining your own interface and using that for stub out unit tests?

Reply:
I didn't want to expand on this too much, but even if the return type is an interface (which `Link` is), the underlying return value is still a struct. `Link` is indeed non-implementable outside of the package, so if you want to stub out a Link in a consuming program, you'd still have to wrap it. I think that's acceptable.